Source/RxSwift/fromAsync.swift

```diff
@@ -66,3 +66,71 @@ extension Observable {
         return { (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) in Observable.fromAsync(curry(asyncRequest)(a)(b)(c)(d)(e)(f)(g)(h)(i)(j)) }
     }
 }
+
+public enum FromAsyncError: Error {
+    /// Both result & error can't be nil
+    case inconsistentCompletionResult
```

Comment:
Is this relevant to Observable fromAsync as well ? Why can't result be optional and nil?

Reply:
Hi @freak4pc . Thank u for ur comments. Do u have any idea how should I handle a situation in which completionHandler have both error and success? I think throwing an inconsistent state error is the best option.

Reply:
If user want to make result optional type, fromAsync generic type would be double optional
`A??`, so it will work ok

Reply:
I mean, outer optional goes for result optionality and inner optional is user defined type

Reply:
I wouldn't treat empty result as an error nevertheless.

Reply:
Why? This is the way, how single works. It should fire exactly one event OR one error. I thunk throwing this documented error would be correct.
I agree about other comments, but here I am pretty sure in implementation

Reply:
Do u have any suggestions? Maybe better way to implement that?

Reply:
@mkko I'm not too familiar with this operator, so if you've used it and have some opinions it would be highly helpful.

Reply:
Sorry for the delay. I think that's a fair point, it does look like SingleAsync also expects a value and conformity should be prioritized.

I'm just thinking about a use case where you're not expecting a value, health/auth checks and such. Does it work if you define a `Single` with an optional value?

Reply:
I fully understand your concerns. But completion handler which has optional value should look like that:
```
typealias ResulType = String? 
func foo(completionHandler: (ResultType?, TestError?) -> Void)
// which is equal to
func foo(completionHandler: (String??, TestError?) -> Void)
```
We use outer optional to show that completion can be called with result OR with an error. And our inner optional goes for our actual result (which type is `String?`l).
About auth check - user should use `Completable` instead. But if he wants to use `Single` its `ElementType` type should be `Void`

Reply:
Fair point. I think this looks good 👍

Reply:
That makes sense - could we add a quick unit test for the optional case, just for sanity and coverage? I'll merge after that.

Reply:
Just added additional test case with optional result type